README
NAME
    Class::Factory::Enhanced - More functionality for Class::Factory

VERSION
    version 1.100810

SYNOPSIS
        package My::Factory;
        use base 'Class::Factory::Enhanced';

        package Some::Class;
        My::Factory->add_factory_type(
            person_name    => 'Person::Name',
            person_address => 'Person::Address'
        );

DESCRIPTION
    This class subclasses Class::Factory and adds some functionality.

METHODS
  add_factory_type
    Like "Class::Factory"'s "add_factory_type()", but this one can add
    several mappings at once. See the Synopsis for an example.

  register_factory_type
    Like "Class::Factory"'s "register_factory_type()", but this one can add
    several mappings at once.

  make_object_for_type
        $factory->make_object_for_type('person_name',
            last_name  => 'Shindou',
            first_name => 'Hikaru',
        );

    An alternative constructor that gets the class to be constructed from
    the factory, then calls "new()" on that class.

    Takes as arguments the object type to be constructed and a list of
    arguments to be passed to the constructor ("new()") of the newly
    constructed object.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Factory-Enhanced
    >.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Class-Factory-Enhanced/>.

    The development version lives at
    <http://github.com/hanekomu/Class-Factory-Enhanced/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2007 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Class-Factory-Enhanced

1.100810  2010-03-22 22:42:55 Europe/Vienna
    - converted distribution to Dist::Zilla-style

0.09  Thu 2010.02.18 21:42:11 CET (Marcel Gruenauer <marcel@cpan.org>)
     - tests no longer require Class::Accessor::Complex

0.08  Mon 2010.02.08 14:40:08 CET (Marcel Gruenauer <marcel@cpan.org>)
     - Changed dist style and Changes back to standard. People didn't like it -
       the nail that sticks out gets hammered down.
     - Added author tests; this also removes dependencies

0.07  Thu, 29 May 2008 12:01:33 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - set the version to 0.07
     - Changes: fixed global.name
     - .shipit: fixed svk.tagpattern

0.06  Wed, 21 Nov 2007 18:21:11 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - Added requirement for Class::Accessor::Installer
     - .shipit: added Twitter step
     - Makefile.PL: added auto_install() and process_templates()
     - lib/*: converted to template
     - updated MANIFEST
     - updated t/perlcriticrc
     - tags: BUGFIX

0.05  Wed, 07 Nov 2007 14:40:43 +0100 (Marcel Gruenauer <marcel@cpan.org>)
     - Makefile.PL now requires Class::Factory 1.06 or better specifically.

0.04  Wed, 07 Nov 2007 08:03:33 +0100 (Marcel Gruenauer <marcel@cpan.org>)
     - removed Class::Factory::Patched, since Fred Moyers accepted my patches
       to Class::Factory. Class::Factory::Enhanced now subclasses
       Class::Factory.

0.03  Wed, 07 Nov 2007 01:07:39 +0100 (Marcel Gruenauer <marcel@cpan.org>)
     - added get_factory_type_for()

0.02  Sun, 21 Oct 2007 21:50:03 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - build_requires 'Class::Accessor::Complex'

0.01  Sun, 21 Oct 2007 20:03:24 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


t/lib/MyFactory.pm
package MyFactory;
our $VERSION = '1.100810';
use warnings;
use strict;
use base 'Class::Factory::Enhanced';

# Default mappings
__PACKAGE__->register_factory_type(
    person_name    => 'Person::SimpleName',
    person_address => 'Person::SimpleAddress',
);
1;


t/lib/Person/Base.pm
package Person::Base;
our $VERSION = '1.100810';
use warnings;
use strict;

# This constructor and the accessors in the subclasses were taken from
# Class::Accessor::Complex. I didn't use Class::Accessor::Complex, however, in
# order not to burden the user with additional requirements when it was easy
# to avoid them.

sub new {
    my $class = shift;
    my $self = ref($class) ? $class : bless {}, $class;
    my %args =
      (scalar(@_ == 1) && ref($_[0]) eq 'HASH')
      ? %{ $_[0] }
      : @_;
    $self->$_($args{$_}) for keys %args;
    $self->init(%args) if $self->can('init');
    $self;
}

1;


t/lib/Person/SimpleName.pm
package Person::SimpleName;
our $VERSION = '1.100810';
use warnings;
use strict;
use base 'Person::Base';

sub fullname {
    return $_[0]->{fullname} if @_ == 1;
    $_[0]->{fullname} = $_[1];
}
1;


lib/Class/Factory/Patched.pm
package Class::Factory::Patched;

# $Id: Factory.pm 40 2006-08-02 05:51:40Z cwinters $

use strict;

$Class::Factory::Patched::VERSION = '1.05';

my %CLASS_BY_FACTORY_AND_TYPE  = ();
my %FACTORY_INFO_BY_CLASS      = ();
my %REGISTER                   = ();

# Simple constructor -- override as needed

sub new {
    my ( $pkg, $type, @params ) = @_;
    my $class = $pkg->get_factory_class( $type );
    return undef unless ( $class );
    my $self = bless( {}, $class );
    return $self->init( @params );
}


# Subclasses should override, but if they don't they shouldn't be
# penalized...

sub init { return $_[0] }

# Find the class associated with $object_type

sub get_factory_class {
    my ( $item, $object_type ) = @_;
    my $class = ref $item || $item;
    my $factory_class =
        $CLASS_BY_FACTORY_AND_TYPE{ $class }->{ $object_type };
    return $factory_class if ( $factory_class );

    $factory_class = $REGISTER{ $class }->{ $object_type };
    if ( $factory_class ) {
        my $added_class =
            $class->add_factory_type( $object_type, $factory_class );
        return $added_class;
    }
    $item->factory_error( "Factory type '$object_type' is not defined ",
                          "in '$class'" );
    return undef;
}


# Associate $object_type with $object_class

sub add_factory_type {
    my ( $item, $object_type, $object_class ) = @_;
    my $class = ref $item || $item;
    unless ( $object_type )  {
        $item->factory_error( "Cannot add factory type to '$class': no ",
                              "type defined");
    }
    unless ( $object_class ) {
        $item->factory_error( "Cannot add factory type '$object_type' to ",
                              "'$class': no class defined" );
    }

    my $set_object_class =
        $CLASS_BY_FACTORY_AND_TYPE{ $class }->{ $object_type };
    if ( $set_object_class ) {
        $item->factory_log( "Attempt to add type '$object_type' to '$class' ",
                            "redundant; type already exists with class ",
                            "'$set_object_class'" );
        return;
    }

    # Make sure the object class looks like a perl module/script
    # Acceptable formats:
    #   Module.pm  Module.ph  Module.pl  Module
    $object_class =~ m/^([\w:-]+(?:\.(?:pm|ph|pl))?)$/;
    $object_class = $1;

    if ( $INC{ $object_class } ) {
        $item->factory_log( "Looks like class '$object_class' was already ",
                            "included; no further work necessary" );
    }
    else {
        eval "require $object_class";
        if ( $@ ) {
            $item->factory_error( "Cannot add factory type '$object_type' to ",
                                  "class '$class': factory class '$object_class' ",
                                  "cannot be required: $@" );
            return undef;
        }
    }

    # keep track of what classes have been included so far...
    $CLASS_BY_FACTORY_AND_TYPE{ $class }->{ $object_type } = $object_class;

    # keep track of what factory and type are associated with a loaded
    # class...
    $FACTORY_INFO_BY_CLASS{ $object_class } = [ $class, $object_type ];

    return $object_class;
}

sub register_factory_type {
    my ( $item, $object_type, $object_class ) = @_;
    my $class = ref $item || $item;
    unless ( $object_type )  {
        $item->factory_error( "Cannot add factory type to '$class': no type ",
                              "defined" );
    }
    unless ( $object_class ) {
        $item->factory_error( "Cannot add factory type '$object_type' to ",
                              "'$class': no class defined" );
    }

    my $set_object_class = $REGISTER{ $class }->{ $object_type };
    if ( $set_object_class ) {
        $item->factory_log( "Attempt to register type '$object_type' with ",
                            "'$class' is redundant; type registered with ",
                            "class '$set_object_class'" );
        return;
    }
    return $REGISTER{ $class }->{ $object_type } = $object_class;
}


sub get_loaded_classes {
    my ( $item ) = @_;
    my $class = ref $item || $item;
    return () unless ( ref $CLASS_BY_FACTORY_AND_TYPE{ $class } eq 'HASH' );
    return sort values %{ $CLASS_BY_FACTORY_AND_TYPE{ $class } };
}

sub get_loaded_types {
    my ( $item ) = @_;
    my $class = ref $item || $item;
    return () unless ( ref $CLASS_BY_FACTORY_AND_TYPE{ $class } eq 'HASH' );
    return sort keys %{ $CLASS_BY_FACTORY_AND_TYPE{ $class } };
}

sub get_registered_classes {
    my ( $item ) = @_;
    my $class = ref $item || $item;
    return () unless ( ref $REGISTER{ $class } eq 'HASH' );
    return sort values %{ $REGISTER{ $class } };
}

sub get_registered_class {
	my ( $item, $type ) = @_;
	unless ( $type ) {
	    warn("No factory type passed");
		return undef;
	}
    my $class = ref $item || $item;
    return undef unless ( ref $REGISTER{ $class } eq 'HASH' );
	return $REGISTER{ $class }{ $type };
}

sub get_registered_types {
    my ( $item ) = @_;
    my $class = ref $item || $item;
    return () unless ( ref $REGISTER{ $class } eq 'HASH' );
    return sort keys %{ $REGISTER{ $class } };
}

# Return the factory class that created $item (which can be an object
# or class)

sub get_my_factory {
    my ( $item ) = @_;
    my $impl_class = ref( $item ) || $item;
    my $impl_info = $FACTORY_INFO_BY_CLASS{ $impl_class };
    if ( ref( $impl_info ) eq 'ARRAY' ) {
        return $impl_info->[0];
    }
    return undef;
}

# Return the type that the factory used to create $item (which can be
# an object or class)

sub get_my_factory_type {
    my ( $item ) = @_;
    my $impl_class = ref( $item ) || $item;
    my $impl_info = $FACTORY_INFO_BY_CLASS{ $impl_class };
    if ( ref( $impl_info ) eq 'ARRAY' ) {
        return $impl_info->[1];
    }
    return undef;
}

########################################
# Overridable Log / Error

sub factory_log   { shift; warn @_, "\n" }
sub factory_error { shift; die @_, "\n" }


# BEGIN PATCH

sub remove_factory_type {
    my ( $item, @object_types ) = @_;
    my $class = ref $item || $item;

    for my $object_type (@object_types) {
        unless ( $object_type )  {
            $item->factory_error(
                "Cannot remove factory type from '$class': no type defined"
            );
        }

        delete $CLASS_BY_FACTORY_AND_TYPE{ $class }->{ $object_type };
    }
}

sub unregister_factory_type {
    my ( $item, @object_types ) = @_;
    my $class = ref $item || $item;

    for my $object_type (@object_types) {
        unless ( $object_type )  {
            $item->factory_error(
                "Cannot remove factory type from '$class': no type defined"
            );
        }

        delete $REGISTER{ $class }->{ $object_type };

        # Also delete from $CLASS_BY_FACTORY_AND_TYPE because if the object
        # type has already been instantiated, then it will have been processed
        # by add_factory_type(), thus creating an entry in
        # $CLASS_BY_FACTORY_AND_TYPE. We can call register_factory_type()
        # again, but when we try to instantiate an object via
        # get_factory_class(), it will find the old entry in
        # $CLASS_BY_FACTORY_AND_TYPE and use that.

        delete $CLASS_BY_FACTORY_AND_TYPE{ $class }->{ $object_type };
    }
}


sub get_factory_type_for {
    my ( $self, $item ) = @_;
    my $impl_class = ref( $item ) || $item;
    my $impl_info = $FACTORY_INFO_BY_CLASS{ $impl_class };
    if ( ref( $impl_info ) eq 'ARRAY' ) {
        return $impl_info->[1];
    }
    return undef;
}


# END PATCH


1;

__END__

=head1 NAME

Class::Factory::Patched - Patched version of Class::Factory

=head1 SYNOPSIS

See L<Class::Factory>

=head1 DESCRIPTION

This is a patched version of L<Class::Factory>. It is included because
Class::Factory has closures over lexical variables, and those variables cannot
be accessed from the outside. It also doesn't allow to add or register
mappings that have already been defined.

However, I need to override mappings. You can see an example in the tests of
L<Class::Accessor::FactoryTyped>, but basically the idea is this:

A person object has a name object and an address object. This person object is
used in two different applications. One application wants the person to just
have a single fullname string and a single address string. The other
application wants the name to be split into first name and last name, and the
address to be split into street address, postal code, city and country.

The second application needs to inherit from the first application because it
shares a lot of common characteristics with it.

We would still like to use the same person object, though, because it
interacts with other parts of the application in some standard way.

So the first application would tell the factory to map the C<person_name> to a
class that just has a simple fullname string and the C<person_address> to a
class that just has a simple address string.

The second application, because it inherits from the first application, also
inherits the factory, but redefines the C<person_name> and C<person_address>
object types to point to the more complex implementations.

This isn't possible with Class::Factory alone, because when the second
application tries to redefine the mappings, Class::Factory doesn't allow it.

Therefore I propose the following additions to Class::Factory. If and when the
author of Class::Factory includes them in Class::Factory, this patch module
will go away.

=head1 PATCHES

=over 4

=item remove_factory_type

Takes a list of object types and removes them from the factory. This is the
opposite of C<add_factory_type()>.

=item unregister_factory_type

Takes a list of object types and unregisters them from the factory. This is
the opposite of C<register_factory_type()>.

=back

=head1 TAGS

If you talk about this module in blogs, on del.icio.us or anywhere else,
please use the C<classfactoryenhanced> tag.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-class-factory-enhanced@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a CPAN site
near you. Or see <http://www.perl.com/CPAN/authors/id/M/MA/MARCEL/>.

=head1 AUTHOR

Marcel GrE<uuml>nauer, C<< <marcel@cpan.org> >>

=head1 COPYRIGHT AND LICENSE

Copyright 2007 by Marcel GrE<uuml>nauer

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut



lib/Class/Factory/Enhanced.pm
use 5.006;
use warnings;
use strict;

package Class::Factory::Enhanced;
our $VERSION = '1.100810';
# ABSTRACT: More functionality for Class::Factory

use base 'Class::Factory';

# add support for defining several mappings at once
sub add_factory_type {
    my ($item, @args) = @_;
    my $class = ref $item || $item;

    # SUPER::add_factory_type, which only allows one mapping, returns the
    # object class. The return value is not used when defining the mappings,
    # but it is used in get_factory_class(). get_factory_class() only calls
    # this with one argument, so return the first object class we see. It's a
    # hack, yes...
    my $first_object_class;
    while (my ($object_type, $object_class) = splice @args, 0, 2) {
        $class->SUPER::remove_factory_type($object_type);
        $class->SUPER::add_factory_type($object_type, $object_class);
        $first_object_class ||= $object_class;
    }
    return $first_object_class;
}

sub register_factory_type {
    my ($item, @args) = @_;
    my $class = ref $item || $item;
    while (my ($object_type, $object_class) = splice @args, 0, 2) {
        $class->SUPER::unregister_factory_type($object_type);
        $class->SUPER::register_factory_type($object_type, $object_class);
    }
}

sub make_object_for_type {
    my ($self, $object_type, @args) = @_;
    my $class = $self->get_factory_class($object_type);
    $class->new(@args);
}
1;


__END__
=pod

=head1 NAME

Class::Factory::Enhanced - More functionality for Class::Factory

=head1 VERSION

version 1.100810

=head1 SYNOPSIS

    package My::Factory;
    use base 'Class::Factory::Enhanced';

    package Some::Class;
    My::Factory->add_factory_type(
        person_name    => 'Person::Name',
        person_address => 'Person::Address'
    );

=head1 DESCRIPTION

This class subclasses L<Class::Factory> and adds some functionality.

=head1 METHODS

=head2 add_factory_type

Like C<Class::Factory>'s C<add_factory_type()>, but this one can add several
mappings at once. See the Synopsis for an example.

=head2 register_factory_type

Like C<Class::Factory>'s C<register_factory_type()>, but this one can add
several mappings at once.

=head2 make_object_for_type

    $factory->make_object_for_type('person_name',
        last_name  => 'Shindou',
        first_name => 'Hikaru',
    );

An alternative constructor that gets the class to be constructed from the
factory, then calls C<new()> on that class.

Takes as arguments the object type to be constructed and a list of arguments
to be passed to the constructor (C<new()>) of the newly constructed object.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Factory-Enhanced>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Class-Factory-Enhanced/>.

The development version lives at
L<http://github.com/hanekomu/Class-Factory-Enhanced/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



inc/Module/Install/StandardTests.pm
#line 1
package Module::Install::StandardTests;

use warnings;
use strict;
use File::Spec;


our $VERSION = '0.05';


use base 'Module::Install::Base';


sub use_standard_tests {
    my ($self, %specs) = @_;
    
    my %with = map { $_ => 1 } qw/compile pod pod_coverage perl_critic/;
    if (exists $specs{without}) {
        $specs{without} = [ $specs{without} ] unless ref $specs{without};
        delete $with{$_} for @{ $specs{without} };
    }

    $self->build_requires('Test::More');
    $self->build_requires('UNIVERSAL::require');

    # Unlike other tests, this is mandatory.
    $self->build_requires('Test::Compile');

    $self->write_standard_test_compile;    # no if; this is mandatory
    $self->write_standard_test_pod          if $with{pod};
    $self->write_standard_test_pod_coverage if $with{pod_coverage};
    $self->write_standard_test_perl_critic  if $with{perl_critic};
}


sub write_test_file {
    my ($self, $filename, $code) = @_;
    $filename = File::Spec->catfile('t', $filename);

    # Outdent the code somewhat. Remove first empty line, if any. Then
    # determine the indent of the first line. Throw that amount of indenting
    # away from any line. This allows you to indent the code so it's visually
    # clearer (see methods below) while creating output that's indented more
    # or less correctly. Smoke result HTML pages link to the .t files, so it
    # looks neater.

    $code =~ s/^ *\n//;
    (my $indent = $code) =~ s/^( *).*/$1/s;
    $code =~ s/^$indent//gm;

    print "Creating $filename\n";
    open(my $fh, ">$filename") or die "can't create $filename $!";

    my $perl = $^X;
    print $fh <<TEST;
#!$perl -w

use strict;
use warnings;

$code
TEST

    close $fh or die "can't close $filename $!\n";
    $self->realclean_files($filename);
}


sub write_standard_test_compile {
    my $self = shift;
    $self->write_test_file('000_standard__compile.t', q/
        BEGIN {
            use Test::More;
            eval "use Test::Compile";
            Test::More->builder->BAIL_OUT(
                "Test::Compile required for testing compilation") if $@;
            all_pm_files_ok();
        }
    /);
}


sub write_standard_test_pod {
    my $self = shift;
    $self->write_test_file('000_standard__pod.t', q/
        use Test::More;
        eval "use Test::Pod";
        plan skip_all => "Test::Pod required for testing POD" if $@;
        all_pod_files_ok();
    /);
}


sub write_standard_test_pod_coverage {
    my $self = shift;
    $self->write_test_file('000_standard__pod_coverage.t', q/
        use Test::More;
        eval "use Test::Pod::Coverage";
        plan skip_all =>
            "Test::Pod::Coverage required for testing POD coverage" if $@;
        all_pod_coverage_ok();
    /);
}


sub write_standard_test_perl_critic {
    my $self = shift;
    $self->write_test_file('000_standard__perl_critic.t', q/
        use FindBin '$Bin';
        use File::Spec;
        use UNIVERSAL::require;
        use Test::More;

        plan skip_all =>
            'Author test. Set $ENV{TEST_AUTHOR} to a true value to run.'
            unless $ENV{TEST_AUTHOR};

        my %opt;
        my $rc_file = File::Spec->catfile($Bin, 'perlcriticrc');
        $opt{'-profile'} = $rc_file if -r $rc_file;

        if (Perl::Critic->require('1.078') &&
            Test::Perl::Critic->require &&
            Test::Perl::Critic->import(%opt)) {

            all_critic_ok("lib");
        } else {
            plan skip_all => $@;
        }
    /);
}


1;

__END__

#line 250



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.006; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Class::Factory::Enhanced',
                       'DISTNAME' => 'Class-Factory-Enhanced',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'More functionality for Class::Factory',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100810',
                       'PREREQ_PM' => {
                                        'Scalar::Util' => '0',
                                        'English' => '0',
                                        'FindBin' => '0',
                                        'Class::Factory' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'File::Find' => '0',
                                        'File::Temp' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



